work_queue/src/work_queue.c

```diff
@@ -851,31 +858,88 @@ static int get_transfer_wait_time(struct work_queue *q, struct work_queue_worker
 	return timeout;
 }
 
-static void remove_factory(struct work_queue_factory_info *f)
+static int factory_trim_workers(struct work_queue *q, struct work_queue_factory_info *f)
 {
-	free(f->name);
-	free(f);
+	char *factory_name;
+	if ( !f || !(factory_name = f->name) ) return 0;
+
+	// Iterate through all workers and shut idle ones down
+	struct work_queue_worker *w;
+	char *key;
+	int trimmed_workers = 0;
+
+	struct hash_table *idle_workers = hash_table_create(0, 0);
+	hash_table_firstkey(q->worker_table);
+	while ( f->connected_workers - trimmed_workers > f->max_workers &&
+			hash_table_nextkey(q->worker_table, &key, (void **) &w) ) {
+		if ( w->factory_name &&
+				!strcmp(factory_name, w->factory_name) &&
+				itable_size(w->current_tasks) < 1 ) {
+			hash_table_insert(idle_workers, key, w);
+			trimmed_workers++;
+		}
+	}
+
+	hash_table_firstkey(idle_workers);
+	while (hash_table_nextkey(idle_workers, &key, (void **) &w)) {
+		hash_table_remove(idle_workers, key);
+		hash_table_firstkey(idle_workers);
+		shut_down_worker(q, w);
+	}
+	hash_table_delete(idle_workers);
+
+	debug(D_WQ, "Trimmed %d workers from %s", trimmed_workers, factory_name);
+	return trimmed_workers;
 }
 
-static void update_factory(struct work_queue *q, struct jx *j)
+static struct work_queue_factory_info *create_factory_info(struct work_queue *q, const char *name)
 {
-	const char *name = jx_lookup_string(j, "factory_name");
 	struct work_queue_factory_info *f;
-
-	if ( !(f = hash_table_lookup(q->factory_table, name)) ) {
-		// Create a new factory
-		debug(D_WQ, "new factory %s detected", name);
-		f = malloc(sizeof(*f));
-		if (!f) {
-			debug(D_NOTICE, "Cannot allocate memory for factory %s.", name);
-			return;
-		}
+	if ( (f = hash_table_lookup(q->factory_table, name)) ) return f;
+	if ( (f = malloc(sizeof(*f))) ) {
```

Comment:
No need to check malloc here. We don't check anywhere else, and if malloc fails, it is not clear what to do anyway.